app/main/views/frameworks.py

```diff
@@ -109,6 +132,53 @@ def update_framework(framework_slug):
             enforce_required=True
         )
 
+    if json_payload['frameworks'].get('status', None) == 'pending' and framework.status != 'pending':
+        # When we set a framework to pending, we want to polyfill all supplier declarations (regardless of whether
+        # their application is 'complete' or not) with information from their supplier account. We have a set of
+        # hardcoded keys for these values. Currently these keys don't need to mean anything outside of the API, so
+        # they are defined in dm-api/app/constants. At the time this was written, it didn't seem feasible to manage
+        # this query through SQLAlchemy in an efficient (i.e. 100% server-side) manner, so we are using raw SQL.
+
+        # We pump this data into the declaration before updating the framework state, so that if for some reason the
+        # framework state fails to update, when a second attempt is made the data is overwritten.
+        statement = text(f"""
+UPDATE
+    supplier_frameworks sf
+SET
+    declaration = sf.declaration::jsonb || (
+        SELECT
+            row_to_json(data)
+        FROM
+        (
+            SELECT
+                s.name AS "{KEY_TRADING_NAME}",
+                s.registered_name AS "{KEY_REGISTERED_NAME}",
+                COALESCE(s.companies_house_number,
+                         s.other_company_registration_number) AS "{KEY_REGISTRATION_NUMBER}",
+                s.registration_date AS "{KEY_REGISTRATION_DATE}",
+                s.duns_number AS "{KEY_DUNS_NUMBER}",
+                s.vat_number AS "{KEY_VAT_NUMBER}",
+                s.trading_status AS "{KEY_TRADING_STATUS}",
+                s.organisation_size AS "{KEY_ORGANISATION_SIZE}",
+                s.registration_country AS "{KEY_REGISTRATION_COUNTRY}",
+                ci.address1 AS "{KEY_REGISTRATION_BUILDING}",
+                ci.city AS "{KEY_REGISTRATION_TOWN}",
+                ci.postcode AS "{KEY_REGISTRATION_POSTCODE}"
+            FROM
+                suppliers AS s
+            JOIN
+                contact_information AS ci ON s.supplier_id = ci.supplier_id
+            WHERE
+                s.supplier_id = sf.supplier_id
+            ORDER BY ci.id ASC
+            LIMIT 1
+        ) AS data
+    )::jsonb
+WHERE
+    sf.framework_id = :framework_id""")
```

Comment:
I would ❤️ love ❤️  for this big old statement (and the excellent comment above it) to be pulled into a helper function for readability. 

It's a real shame there's no ORM equivalent (firebreak project anyone?) of the JSON update, but given the expensiveness of the operation, raw SQL seems the sensible way to go.

Reply:
I'm going to keep trying to work out how this could be ORM-ified. If I still haven't done it by next firebreak, then yeah, I'll use that time to have another serious crack at it.